stp_zmq/client_message_provider.py

```diff
@@ -0,0 +1,99 @@
+from collections import OrderedDict
+from typing import Tuple, Optional
+
+import zmq
+
+from plenum.common.exceptions import InvalidMessageExceedingSizeException
+from plenum.common.timer import QueueTimer, RepeatingTimer
+from stp_core.common.constants import CONNECTION_PREFIX
+from stp_core.common.log import getlogger
+
+logger = getlogger()
+
+
+class ClientMessageProvider:
+    def __init__(self, name, config, prepare_to_send, metrics, mt_outgoing_size,
+                 timer=None, listener=None):
+        self._name = name
+        self.metrics = metrics
+        self.listener = listener
+        self._prepare_to_send = prepare_to_send
+        self._mt_outgoing_size = mt_outgoing_size
+        self._config = config
+        self._timer = QueueTimer() if timer is None else timer
+        self._pending_client_messages = OrderedDict()
+        RepeatingTimer(self._timer, self._config.RESEND_CLIENT_MSG_TIMEOUT, self._send_pending_messages)
+        RepeatingTimer(self._timer, self._config.REMOVE_CLIENT_MSG_TIMEOUT, self._remove_old_messages)
+
+    def transmit_through_listener(self, msg, ident) -> Tuple[bool, Optional[str]]:
+        self._pending_client_messages.setdefault(ident, []).append((self._timer.get_current_time(),
```

Comment:
So, we do this for every message, even if it's successfully  transmitted and will be removed from the list at the end of the method. Since absolute majority of requests will be transmitted successfully, should we do this stashing only in case transmit failed?